### PR TITLE
Optimize timer

### PR DESCRIPTION
**Pull request by ces42**
This PR deals with the timer mechanism. Producing timer events every 0.3 second, adds up to a small but avoidable use of resources over time.
Meanwhile the only reason the current code needs frequent timer events is to trigger hold gestures. So the solution is to have a very infrequent timer (I set it to every 5 seconds so that expired events get cleared eventually) and schedule extra timer events to check hold gestures.
As added benefits this will 
 - make hold gesture detection more accurate (instead of being off by possibly up to 0.3 seconds)
 - moves the timer from a separate process into a thread thereby reducing memory usage

One caveat: Any other plugins that rely on frequent timer events will have to be rewritten to request timer events with `Timer.wake_early`.

### CODE REVIEW

File: lib/fusuma/plugin/inputs/timer_input.rb [hunk 1]
```diff
@@ -26,20 +35,25 @@ def io
...
-              sleep interval
-              writer.puts "timer"
+              begin
+                Timeout.timeout(next_wake - Time.now) do
+                  next_wake = [@early_wake_queue.deq, next_wake].min
+                end
+              rescue Timeout::Error
+                writer.puts "timer"
+                next_wake = Time.now + delta_t
+              end
```
Comment (iberianpig) [inline lib/fusuma/plugin/inputs/timer_input.rb:56]: Interesting. 
- Blocking with dequeue.
- Reduce timeout value with dequeued
- Send a timer event on timeout.

File: lib/fusuma.rb [hunk 2]
```diff
@@ -56,7 +56,10 @@ def load_custom_config(config_path = nil)
     end
 
     def initialize
-      @inputs = Plugin::Inputs::Input.plugins.map(&:new)
+      @inputs = Plugin::Inputs::Input.plugins.map do |cls|
+        cls.ancestors.include?(Singleton) ? cls.instance : cls.new
+      end
+      @inputs.push(Plugin::Inputs::TimerInput.instance)
```
Comment (iberianpig) [inline lib/fusuma.rb:62]: Why does the instance of TimerInput add again to the `@inputs` array?

**Comment by ces42** [comment]
Oh, I think that's a mistake.

**Comment by ces42** [comment]
fixed now

**Review by iberianpig — APPROVED** [review]

**Comment by iberianpig** [comment]
Thanks for your contribution.

**Pull request merged**